File: mobile/src/utils/route.py
import os
import math
import random

from typing import List

from kivy.app import App

import gpxpy
import gpxpy.gpx
from geopy.distance import geodesic
from shapely.geometry import Point, LineString


from utils.poi import Poi, PoiType

GPX_ROUTE_SUBDIR="gpx_routes"


class Route:

    @classmethod
    def load_gpx(cls, filename):
        data_directory = App.get_running_app().user_data_dir
        file_path = os.path.join(data_directory,GPX_ROUTE_SUBDIR, filename)

        with open(file_path, 'r') as gpx_file:
            gpx = gpxpy.parse(gpx_file)

        pois = []
        for track in gpx.tracks:
            for segment in track.segments:
                for point in segment.points:
                    poi = Poi(lat=point.latitude, lon=point.longitude, name=point.name, type=PoiType.ROUTE)
                    pois.append(poi)
        return Route(name=filename, points=pois)



    def __init__(self, name, points: List[Poi]):
        self.name = name
        self.points = points



    def dump_gpx(self):
        data_directory = App.get_running_app().user_data_dir
        gpx_directory = os.path.join(data_directory, GPX_ROUTE_SUBDIR)
        file_path = os.path.join(gpx_directory, self.name)

        # Ensure the directory exists
        os.makedirs(gpx_directory, exist_ok=True)

        gpx = gpxpy.gpx.GPX()

        # Create track
        gpx_track = gpxpy.gpx.GPXTrack(name=self.name)
        gpx.tracks.append(gpx_track)

        # Create segment
        gpx_segment = gpxpy.gpx.GPXTrackSegment()
        gpx_track.segments.append(gpx_segment)

        # Add points to segment
        for poi in self.points:
            if poi.type == PoiType.ROUTE:
                gpx_point = gpxpy.gpx.GPXTrackPoint(poi.lat, poi.lon, name=poi.name)
                gpx_segment.points.append(gpx_point)

        xml = gpx.to_xml()
        with open(file_path, 'w') as gpx_file:
            gpx_file.write(xml)

        print(f"GPX data written to {file_path}")


    def calculate_waypoint_route(self) -> Route:
        min_segment_length = 15
        distance_to_insert = 10

        modified_points = []

        for i, current_poi in enumerate(self.points):
            prev_poi, next_poi = self.find_adjacent_poi(current_poi)

            # Add extra previous point
            if prev_poi and self.is_special_coordinate(current_poi):
                prev_segment_length = geodesic((current_poi.lat, current_poi.lon), (prev_poi.lat, prev_poi.lon)).meters
                if prev_segment_length > min_segment_length:
                    # Calculate and insert extra previous point
                    extra_prev_poi = self.interpolate_poi(current_poi, prev_poi, distance_to_insert)
                    modified_points.append(extra_prev_poi)

            # Add the current point
            modified_points.append(current_poi)

            # Add extra next point
            if next_poi and self.is_special_coordinate(current_poi):
                next_segment_length = geodesic((current_poi.lat, current_poi.lon), (next_poi.lat, next_poi.lon)).meters
                if next_segment_length > min_segment_length:
                    # Calculate and insert extra next point
                    extra_next_poi = self.interpolate_poi(current_poi, next_poi, distance_to_insert)
                    modified_points.append(extra_next_poi)

        return Route(name = self.name, points=modified_points)


    def find_adjacent_poi(self, poi) -> (Poi, Poi):
        """Find adjacent points (previous and next) for a given Poi in the route."""
        try:
            current_index = self.points.index(poi)
            prev_poi = self.points[current_index - 1] if current_index > 0 else None
            next_poi = self.points[current_index + 1] if current_index < len(self.points) - 1 else None
            return prev_poi, next_poi
        except ValueError:
            # Current poi not in self.points
            return None, None


    def calculate_angle(self, p2):
        """Calculates the angle of p2. p2 is in between of p1 and p3 which are determined from the route waypoints.
        """
        p1, p3 = self.find_adjacent_poi(p2)
        # Check if all points are given
        if not p1 or not p2 or not p3:
            return 0

        # Calculate distances between the points
        a = geodesic((p2.lat, p2.lon), (p3.lat, p3.lon)).meters
        b = geodesic((p1.lat, p1.lon), (p3.lat, p3.lon)).meters
        c = geodesic((p1.lat, p1.lon), (p2.lat, p2.lon)).meters

        # Apply the Law of Cosines to find the angle at p2
        # Clamp cos_angle to the valid range [-1, 1] to avoid math domain error
        cos_angle = max(-1, min(1, (a**2 + c**2 - b**2) / (2 * a * c)))
        angle = math.acos(cos_angle) * (180 / math.pi)

        return angle


    def find_closest_segment(self, poi):
        min_distance = float('inf')
        closest_segment_start_index = None

        for i in range(len(self.points) - 1):
            start_coord = self.points[i]
            end_coord = self.points[i + 1]

            # Create a line segment and a point
            line = LineString([(start_coord.lon, start_coord.lat), (end_coord.lon, end_coord.lat)])
            point = Point(poi.lon, poi.lat)

            # Use Shapely to compute the distance
            current_distance = line.distance(point)

            if current_distance < min_distance:
                min_distance = current_distance
                closest_segment_start_index = i

        if closest_segment_start_index is None:
            raise ValueError("Could not find closest segment.")

        closest_segment_end_index = closest_segment_start_index + 1
        closest_segment = {
            "start": {
                "index": closest_segment_start_index,
                "coord": self.points[closest_segment_start_index]
            },
            "end": {
                "index": closest_segment_end_index,
                "coord": self.points[closest_segment_end_index]
            },
            "distance": min_distance
        }

        return closest_segment


    def get_random_point_near_route(self) -> Poi:
        if not self.points:
            raise ValueError("The route has no points.")

        random_index = random.randint(0, len(self.points) - 1)  # Random index in the route
        random_radius = 2 + random.random() * 18  # Radius between 2 and 20 meters
        random_bearing = random.random() * 360  # Random bearing in degrees

        original_point = self.points[random_index]
        # Calculate the new point using geodesic distance and random bearing
        destination = geodesic(kilometers=random_radius / 1000).destination((original_point.lat, original_point.lon), random_bearing)

        return Poi(lat=destination.latitude,lon=destination.longitude)
    
    
    def is_special_coordinate(self, poi) -> bool:
        """Decides whenever the coordinate is special. Special in the sense of pedestrian navigation. In this case the point needs
        some vertex in front and behind of them.
        """
        angle = self.calculate_angle(poi)
        return 60 <= angle <= 120


    def interpolate_poi(self, from_poi, to_poi, distance) -> Poi:
        # Logic to calculate the position of the extra point
        # This might involve bearing calculation and distance offset
        # Placeholder for the actual implementation
        lat_offset = (to_poi.lat - from_poi.lat) * (distance / geodesic((from_poi.lat, from_poi.lon), (to_poi.lat, to_poi.lon)).meters)
        lon_offset = (to_poi.lon - from_poi.lon) * (distance / geodesic((from_poi.lat, from_poi.lon), (to_poi.lat, to_poi.lon)).meters)
        return Poi(lat=from_poi.lat + lat_offset, lon=from_poi.lon + lon_offset, poi_type=PoiType.SYNTHETIC)

File: app/lib/models/route.dart
import 'dart:convert';
import 'dart:math' as math;
import 'package:candle/utils/geo.dart';
import 'package:collection/collection.dart';
import 'package:flutter/widgets.dart';
import 'package:xml/xml.dart' as xml;

import 'package:candle/models/navigation_point.dart';
import 'package:latlong2/latlong.dart';

class Route {
  int? id;
  String name;
  String annotation;
  List<NavigationPoint> points;

  Route({
    this.id,
    required this.name,
    this.annotation = "",
    required this.points,
  });

  Route calculateWaypointRoute() {
    const double minSegmentLength = 15;
    const double distanceToInsert = 10;

    List<NavigationPoint> modifiedPoints = [];

    for (int i = 0; i < points.length; i++) {
      NavigationPoint currentPoi = points[i];
      List<NavigationPoint?> adjacentPois = findAdjacentPoint(currentPoi);
      NavigationPoint? prevPoi = adjacentPois[0];
      NavigationPoint? nextPoi = adjacentPois[1];

      // Add extra previous point
      if (prevPoi != null && isSpecialCoordinate(currentPoi)) {
        double prevSegmentLength = calculateDistance(currentPoi.coordinate, prevPoi.coordinate);
        if (prevSegmentLength > minSegmentLength) {
          // Calculate and insert extra previous point
          NavigationPoint extraPrevPoi = interpolate(
            start: currentPoi,
            end: prevPoi,
            distance: distanceToInsert,
          );
          modifiedPoints.add(extraPrevPoi);
        }
      }

      // Add the current point
      modifiedPoints.add(currentPoi);

      // Add extra next point
      if (nextPoi != null && isSpecialCoordinate(currentPoi)) {
        double nextSegmentLength = calculateDistance(currentPoi.coordinate, nextPoi.coordinate);
        if (nextSegmentLength > minSegmentLength) {
          // Calculate and insert extra next point
          NavigationPoint extraNextPoi = interpolate(
            start: currentPoi,
            end: nextPoi,
            distance: distanceToInsert,
          );
          modifiedPoints.add(extraNextPoi);
        }
      }
    }

    return Route(name: name, points: modifiedPoints);
  }

  Map<String, dynamic> findClosestSegment(LatLng coord) {
    double minDistance = double.infinity;
    int? closestSegmentStartIndex;

    for (int i = 0; i < points.length - 1; i++) {
      NavigationPoint startCoord = points[i];
      NavigationPoint endCoord = points[i + 1];

      // Calculate the distance from the poi to the line segment
      double currentDistance = distanceToSegment(
        point: coord,
        start: startCoord.coordinate,
        end: endCoord.coordinate,
      );
      if (currentDistance < minDistance) {
        minDistance = currentDistance;
        closestSegmentStartIndex = i;
      }
    }

    if (closestSegmentStartIndex == null) {
      throw ArgumentError("Could not find closest segment.");
    }

    int closestSegmentEndIndex = closestSegmentStartIndex + 1;
    return {
      "distance": minDistance,
      "start": {"index": closestSegmentStartIndex, "coord": points[closestSegmentStartIndex]},
      "end": {"index": closestSegmentEndIndex, "coord": points[closestSegmentEndIndex]},
    };
  }

  double calculateTotalLength() {
    double totalLength = 0.0;

    for (int i = 0; i < points.length - 1; i++) {
      totalLength += calculateDistance(
        points[i].coordinate,
        points[i + 1].coordinate,
      );
    }

    return totalLength;
  }

  double calculateResumingLengthFromWaypoint(NavigationPoint currentWaypoint) {
    int waypointIndex = points.indexOf(currentWaypoint);
    if (waypointIndex == -1 || waypointIndex == points.length - 1) {
      // If the current waypoint is not in the list or it is the last point
      return 0.0;
    }

    double totalDistance = 0.0;
    for (int i = waypointIndex; i < points.length - 1; i++) {
      totalDistance += calculateDistance(points[i].coordinate, points[i + 1].coordinate);
    }

    return totalDistance;
  }

  double calculateAngle(NavigationPoint p2) {
    List<NavigationPoint?> adjacentPoints = findAdjacentPoint(p2);

    // Check if adjacent points exist and the list has exactly two elements
    if (adjacentPoints.length != 2 || adjacentPoints.any((point) => point == null)) {
      return 0.0;
    }

    NavigationPoint p1 = adjacentPoints[0]!;
    NavigationPoint p3 = adjacentPoints[1]!;

    // Calculate distances between the points
    double a = calculateDistance(p2.coordinate, p3.coordinate);
    double b = calculateDistance(p1.coordinate, p3.coordinate);
    double c = calculateDistance(p1.coordinate, p2.coordinate);

    // Apply the Law of Cosines to find the angle at p2
    double cosAngle = (math.pow(a, 2) + math.pow(c, 2) - math.pow(b, 2)) / (2 * a * c);
    cosAngle = math.max(-1, math.min(1, cosAngle));
    double angle = math.acos(cosAngle) * (180 / math.pi);

    return angle;
  }

  bool isSpecialCoordinate(NavigationPoint poi) {
    /// Decides whenever the coordinate is special. Special in the sense of pedestrian navigation.
    /// In this case, the point needs some vertex in front and behind of them.
    double angle = calculateAngle(poi);
    return 60 <= angle && angle <= 120;
  }

  List<NavigationPoint?> findAdjacentPoint(NavigationPoint poi) {
    int currentIdx = points.indexWhere((p) => p == poi);

    NavigationPoint? prevPoi;
    NavigationPoint? nextPoi;

    if (currentIdx != -1) {
      prevPoi = currentIdx > 0 ? points[currentIdx - 1] : null;
      nextPoi = currentIdx < points.length - 1 ? points[currentIdx + 1] : null;
    }

    return [prevPoi, nextPoi]; // Return as a list
  }

  NavigationPoint interpolate({
    required NavigationPoint start,
    required NavigationPoint end,
    required double distance,
    NavigationPointType type = NavigationPointType.syntetic, // Default to synthetic
  }) {
    var from = start.coordinate;
    var to = end.coordinate;
    const Distance calculator = Distance();
    final double totalDistance = calculator.as(
      LengthUnit.Meter,
      start.coordinate,
      end.coordinate,
    );

    final double latOffset = (to.latitude - from.latitude) * (distance / totalDistance);
    final double lonOffset = (to.longitude - from.longitude) * (distance / totalDistance);

    return NavigationPoint(
      coordinate: LatLng(from.latitude + latOffset, from.longitude + lonOffset),
      annotation: "",
      type: type, // Set the type
    );
  }

  String toGPX() {
    var builder = xml.XmlBuilder();
    builder.processing('xml', 'version="1.0" encoding="UTF-8"');
    builder.element('gpx', nest: () {
      builder.attribute('version', '1.1');
      builder.attribute('creator', 'YourAppName'); // Replace with your app name
      builder.element('trk', nest: () {
        builder.element('name', nest: name);
        builder.element('trkseg', nest: () {
          for (var point in points) {
            builder.element('trkpt', nest: () {
              builder.attribute('lat', point.coordinate.latitude.toString());
              builder.attribute('lon', point.coordinate.longitude.toString());
              builder.element('ele', nest: '0'); // Elevation, set to 0 if not available
              builder.element('time',
                  nest: DateTime.now()
                      .toIso8601String()); // Current time, replace if actual time data is available
            });
          }
        });
      });
    });

    return builder.buildDocument().toXmlString(pretty: true);
  }

  // Convert the list of NavigationPoints to a JSON string
  String pointsToJson() => json.encode(points.map((x) => x.toMap()).toList());

  // Convert a JSON string to a list of NavigationPoints
  static List<NavigationPoint> pointsFromJson(String jsonString) =>
      List<NavigationPoint>.from(json.decode(jsonString).map((x) => NavigationPoint.fromMap(x)));

  Route copyWith({
    ValueGetter<int?>? id,
    String? name,
    String? annotation,
    List<NavigationPoint>? points,
  }) {
    return Route(
      id: id?.call() ?? this.id,
      name: name ?? this.name,
      annotation: annotation ?? this.annotation,
      points: points ?? this.points,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'name': name,
      'annotation': annotation,
      'points': points.map((x) => x.toMap()).toList(),
    };
  }

  factory Route.fromMap(Map<String, dynamic> map) {
    return Route(
      id: map['id']?.toInt(),
      name: map['name'] ?? '',
      annotation: map['annotation'] ?? '',
      points: map['points'] != null ? Route.pointsFromJson(map['points']) : [],
    );
  }

  String toJson() => json.encode(toMap());

  factory Route.fromJson(String source) => Route.fromMap(json.decode(source));

  @override
  String toString() {
    return 'Route(id: $id, name: $name, annotation: $annotation, points: $points)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    final listEquals = const DeepCollectionEquality().equals;

    return other is Route &&
        other.id == id &&
        other.name == name &&
        other.annotation == annotation &&
        listEquals(other.points, points);
  }

  @override
  int get hashCode {
    return id.hashCode ^ name.hashCode ^ annotation.hashCode ^ points.hashCode;
  }
}
